File: SenLib/Sen1/FileFixes/scripts_scena_dat_us_t0010_dat.cs
using HyoutaUtils;
using System.Collections.Generic;
using System.IO;

namespace SenLib.Sen1.FileFixes {
	public class scripts_scena_dat_us_t0010_dat : FileMod {
		public string GetDescription() {
			return "Fix formatting issues in Thors (bottom floor).";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var s = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x8a76ff88baf96b5eul, 0x72e675d0d5d3b75aul, 0x72cc3989u));
			if (s == null) {
				return null;
			}
			MemoryStream bin = s.CopyToMemoryAndDispose();

			var patcher = new SenScriptPatcher(bin);
			patcher.RemovePartialCommand(0xfbbf, 0x53, 0xfbcf, 1);

			// formatting issues in Towa Chapter 3 Day bonding event
			bin.SwapBytes(0x292ef, 0x292f4);
			bin.SwapBytes(0x29cf4, 0x29cf8);
			bin.Position = 0x29d48;
			bin.WriteUInt8(0x01);
			bin.SwapBytes(0x29e57, 0x29e5b);
			bin.SwapBytes(0x29e78, 0x29e7e);
			bin.SwapBytes(0x2a13e, 0x2a143);
			bin.SwapBytes(0x2a3fc, 0x2a3ff);
			bin.SwapBytes(0x2a421, 0x2a42a);
			bin.SwapBytes(0x2a6e3, 0x2a6e6);
			bin.SwapBytes(0x2a72a, 0x2a72f);

			// formatting issues in Fie Chapter 4 Day bonding event
			bin.SwapBytes(0x2b045, 0x2b04a);
			bin.SwapBytes(0x2b069, 0x2b070);

			// formatting issues in Fie Chapter 5 Evening bonding event
			bin.SwapBytes(0x2bda3, 0x2bda7);
			bin.SwapBytes(0x2bdc6, 0x2bdcb);
			bin.SwapBytes(0x2c02e, 0x2c032);

			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_us/t0010.dat", bin) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var s = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x8a76ff88baf96b5eul, 0x72e675d0d5d3b75aul, 0x72cc3989u));
			if (s == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_us/t0010.dat", s) };
		}
	}
}

File: native/sen1/file_fixes/scripts_scena_dat_us_t0010_dat.cpp
#include <string_view>
#include <vector>

#include "sen/file_getter.h"
#include "p3a/pack.h"
#include "p3a/structs.h"
#include "sen/sen_script_patcher.h"
#include "sha1.h"
#include "util/memwrite.h"

namespace SenLib::Sen1::FileFixes::scripts_scena_dat_us_t0010_dat {
std::string_view GetDescription() {
    return "Fix formatting issues in Thors (bottom floor).";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    try {
        auto file = getCheckedFile(
            "data/scripts/scena/dat_us/t0010.dat",
            254098,
            SenPatcher::SHA1FromHexString("8a76ff88baf96b5e72e675d0d5d3b75a72cc3989"));
        if (!file) {
            return false;
        }

        auto& bin = file->Data;

        using namespace HyoutaUtils::MemWrite;

        SenScriptPatcher patcher(bin);
        patcher.RemovePartialCommand(0xfbbf, 0x53, 0xfbcf, 1);

        // formatting issues in Towa Chapter 3 Day bonding event
        std::swap(bin[0x292ef], bin[0x292f4]);
        std::swap(bin[0x29cf4], bin[0x29cf8]);
        WriteUInt8(&bin[0x29d48], 0x01);
        std::swap(bin[0x29e57], bin[0x29e5b]);
        std::swap(bin[0x29e78], bin[0x29e7e]);
        std::swap(bin[0x2a13e], bin[0x2a143]);
        std::swap(bin[0x2a3fc], bin[0x2a3ff]);
        std::swap(bin[0x2a421], bin[0x2a42a]);
        std::swap(bin[0x2a6e3], bin[0x2a6e6]);
        std::swap(bin[0x2a72a], bin[0x2a72f]);

        // formatting issues in Fie Chapter 4 Day bonding event
        std::swap(bin[0x2b045], bin[0x2b04a]);
        std::swap(bin[0x2b069], bin[0x2b070]);

        // formatting issues in Fie Chapter 5 Evening bonding event
        std::swap(bin[0x2bda3], bin[0x2bda7]);
        std::swap(bin[0x2bdc6], bin[0x2bdcb]);
        std::swap(bin[0x2c02e], bin[0x2c032]);

        // missing 'to' in Rufus line (voice match, also makes more sense) (Chapter 6 intro)
        patcher.ExtendPartialCommand(0x2585e, 0x48, 0x25883, {{0x20, 0x74, 0x6f}});

        result.emplace_back(std::move(bin), file->Filename, SenPatcher::P3ACompressionType::LZ4);

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen1::FileFixes::scripts_scena_dat_us_t0010_dat
